## Homework 1, CS685 Fall 2021

### This is due on November 5th, 2021. This notebook is to be submitted via Gradescope as a PDF file, while your three dataset files (annotator1.csv, annotator2.csv, and final_data.csv) should be emailed to cs685instructors@gmail.com with the subject line formatted as **Firstname_Lastname_HW1data**. 100 points total.

#### IMPORTANT: After copying this notebook to your Google Drive, please paste a link to it below. To get a publicly-accessible link, hit the *Share* button at the top right, then click "Get shareable link" and copy over the result. If you fail to do this, you will receive no credit for this homework!
***LINK:*** https://colab.research.google.com/drive/1ioBEoNZxk93p-mpUiDAjJH7ZnR6ZJPer?usp=sharing

---


##### *How to submit this problem set:*
- Write all the answers in this Colab notebook. Once you are finished, generate a PDF via (File -> Print -> Save as PDF) and upload it to Gradescope.
  
- **Important:** check your PDF before you submit to Gradescope to make sure it exported correctly. If Colab gets confused about your syntax, it will sometimes terminate the PDF creation routine early.

- **Important:** on Gradescope, please make sure that you tag each page with the corresponding question(s). This makes it significantly easier for our graders to grade submissions, especially with the long outputs of many of these cells. We will take off points for submissions that are not tagged.

- When creating your final version of the PDF to hand in, please do a fresh restart and execute every cell in order. One handy way to do this is by clicking `Runtime -> Run All` in the notebook menu.

---

##### *Academic honesty*

- We will audit the Colab notebooks from a set number of students, chosen at random. The audits will check that the code you wrote actually generates the answers in your PDF. If you turn in correct answers on your PDF without code that actually generates those answers, we will consider this a serious case of cheating. See the course page for honesty policies.

- We will also run automatic checks of Colab notebooks for plagiarism. Copying code from others is also considered a serious case of cheating.

---

# Part 1: Annotation

In this homework, you will first collect a labeled dataset of **120** sentences for a text classification task of your choice. This process will include:

1. *Data collection*: Collect 120 sentences from any source you find interesting (e.g., literature, Tweets, news articles, reviews, etc.)

2. *Task design*: Come up with a binary (i.e., only two labels) sentence-level classification task that you would like to perform on your sentences. Don't choose something boring like sentiment analysis... Be creative! Write up annotator guidelines/instructions on how you would like people to label your data.

3. On your dataset, collect annotations from **two** classmates for your task. Everyone in this class will need to both create their own dataset and also serve as an annotator for two other classmates. In order to get everything done on time, you need to complete the following steps:

> *   Find two classmates willing to label 120 sentences each (we will open a Piazza thread to help facilitate this).
*   Send them your annotation guidelines and a way that they can easily annotate the data (e.g., a spreadsheet or Google form)
*   Collect the labeled data from each of the two annotators.
*   Sanity check the data for basic cleanliness (are all examples annotated? are all labels allowable ones?)

4. Collect feedback from annotators about the task including annotation time and obstacles encountered (e.g., maybe your guidelines were confusing! or maybe some sentences were particularly hard to annotate!)

5. Calculate and report inter-annotator agreement.

6. Aggregate output from both annotators to create final dataset.

7. Perform NLP experiments on your new dataset!

## Question 1.1 (10 points):
Describe the source of your unlabeled data, why you chose it, and what kind of sentence selection process you used (if any) to choose 120 sentences for annotation. Also briefly describe the text classification task that you will be collecting labels for in the next section.

### *WRITE YOUR ANSWER HERE* ###
#### Dataset Description: ####
Gender Classification for Political Tweets  

#### Dataset Source: ####
I pulled my text data from Twitter using a tweet scraping code written by me (https://github.com/shubham-shetty/twitter-scraper). I pulled the 20 latest tweets from multiple political figures, journalists and commentators and manually selected 120 relevant tweets from this corpus.  

#### Task & Motivation: ####
Ideally, politics should see equal representation from both sexes. However, American politics is seen to be largely male-dominated. Political communications between genders may have differences due to this gender-salient context (Hu & Kearney, 2020). These communication differences may also manifest in electronic communication such as on Twitter. It is important that the role of the marginalized sex in American politics has an equal voice on such a platform, and the issues they raise have as much weightage as anything else. Analyzing political tweets and classifying them as Male or Female can help in ensuring that Twitter as a platform provides equal opportunity for heterogeneous voices to be heard, and does not act as a homogeneous echo chamber.

## Question 1.2 (25 points):
Copy the annotation guidelines that you provided to your classmates below. We expect that these guidelines will be very detailed (and as such could be fairly long). You must include:

> *   The two categories for your binary classification problem, including the exact strings you want annotators to use while labeling. 
*   Descriptions of the categories and what they mean.
*   Representative examples of each category (i.e., sentences from outside your dataset that you have manually labeled to give annotators an idea of how to perform the task)
*   A discussion of of tricky corner cases, and criteria to help the annotator decide them. If you look at the data and think about how an annotator could do the task, you will likely find a bunch of these!


---




### *COPY YOUR ANNOTATION GUIDELINES HERE.* Please format them nicely so it's easy for us to read / grade :)### 

1. Tweets will be labeled as either 'M' for male or 'F' for female.  

2. Tweets dealing with topics which are largely considered to be associated with men can be labelled as 'M', for instance defense, gun control, sports etc. Examples -   
    a. Democrats have taken our roads, bridges, ports, airports, and waterways hostage to ram through an historically reckless taxing and spending spree that would hurt families and help China. The far left is running Congress and the American people are hurting. (@LeaderMcConnell)  
    b. Why are people worried about giving IRS more information? By a 72% to 14% margin, voters think some agents or political leaders would use additional information to harm political opponents. (@ScottWRasmussen)

3.  Tweets dealing with topics which are largely considered to be associated with women can be labelled as 'F', for instance - reproductive rights, family law, female representation etc. Examples -  
    a. Happy International #DayoftheGirl! Today I’m celebrating the promise that exists within every girl. When we invest in that promise, we empower them to change their communities for the better. Join me and the @GirlsAlliance in supporting girls’ education (@MichelleObama)  
    b. This is the day I learned my Pelosi bio, MADAM SPEAKER, has sold ZERO copies in Wichita, where I was born and raised. So I ask my brother....my classmates from Kos Harris Elementary &amp; Robinson Jr HS &amp; Southeast HS: Really? Does @Watermarkbooks even stock it? #HeartlandHeartbreak (@SusanPage)  

4. More ambiguous tweets can be labelled based on following observations (Bamman et al., 2014)-  
    a. Females tend to use more of the following text markers - pronouns, emotion terms, CMC words (lol, omg), emoticons, hesitation words (err, umm)  
    b.  Males tend to use more of the following text markers - numbers, technology words, swear words.  
    c.  Following markers have mixed or negligible usage - family terms, assent, negation, articles, quantifiers.  

5. Trust your instincts  


## Question 1.3 (5 points):
Write down the names and emails of the two classmates who will be annotating your data below. Once they are finished annotating, create two .csv files (annotator1.csv and annotator2.csv) that contains each annotator's labels. The file should have two columns with headers **text** and **label**, respectively. You will include these files in an email to the instructors account when you're finished with this homework. 

*The tweets.csv file provided as an example in Part 2 below uses the same format.*

### *WRITE CLASSMATE 1 NAME/EMAIL HERE:* ###   
Adam Viola (aviola@umass.edu)

### *WRITE CLASSMATE 2 NAME/EMAIL HERE:* ###   
Pranjali Parse (pparse@umass.edu)

## Question 1.4 (10 points):
After both annotators have finished labeling the 120 sentences you gave them, ask them for feedback about your task and the provided annotation guidelines. If you were to collect more labeled data for this task in the future, what would you change from your current setup? Why? Please include a summary of annotator feedback (with specific examples that they found challenging to label) in your answer.

### *WRITE ANSWER HERE* ###  
Following points were raised through feedback from the annotators -  

1. Too many tweets were ambiguous.
2. Annotators were vary of bias creeping into their annotations.
3. Is it actually possible to just bifurcate tweets into male and female? How would it handle the case when there are more than two genders?


Some changes which could be made if this annotation task was undertaken at a larger scale, keeping these feedback in mind, could be -   

1. Needs more clearer guidelines for classifying tweets as male or female. More research could be done to understand female-centric and male-centric topics in the political sphere. 
2. Instead of providing just 2 different classes to label, maybe we could provide a set of attributes on which annotators could rate tweets on a scale (like 1-5, 1 for least likely to 5 for most likely).

## Question 1.5 (10 points):
Now, compute the inter-annotator agreement between your two annotators. Upload both .csv files to your Colab session (click the folder icon in the sidebar to the left of the screen). In the code cell below, read the data from the two files and compute both the raw agreement (% of examples for which both annotators agreed on the label) and the [Cohen's Kappa](https://en.wikipedia.org/wiki/Cohen%27s_kappa). Feel free to use implementations in existing libraries (e.g., [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html)). After you're done, paste the numbers in the text cell that follows your code. 

*If you're curious, Cohen suggested the Kappa result be interpreted as follows: values ≤ 0 as indicating no agreement and 0.01–0.20 as none to slight, 0.21–0.40 as fair, 0.41– 0.60 as moderate, 0.61–0.80 as substantial, and 0.81–1.00 as almost perfect agreement.*

In [ ]:
# This mounts your Google Drive to the Colab VM.
from google.colab import drive
drive.mount('/content/gdrive', force_remount=True)

# Define folder for annotated data
FOLDERNAME = 'Coursework/Fall 2021/685/Assignments/annotated_data'
assert FOLDERNAME is not None, "[!] Enter the foldername."

# Now that we've mounted your Drive, this ensures that
# the Python interpreter of the Colab VM can load
# python files from within it.
import sys
sys.path.append('/content/gdrive/My Drive/{}'.format(FOLDERNAME))

#Switch to working directory
%cd /content/gdrive/My\ Drive/$FOLDERNAME

Mounted at /content/gdrive
/content/gdrive/My Drive/Coursework/Fall 2021/685/Assignments/annotated_data


In [ ]:
### WRITE CODE TO LOAD ANNOTATIONS AND 
### COMPUTE AGREEMENT + COHEN'S KAPPA HERE!
import pandas as pd
from sklearn import metrics as skl

# Load annotations
annotator1 = pd.read_csv('annotator1.csv', header=0, names=['row_num','text','gender'], usecols = ['text','gender'])
annotator2 = pd.read_csv('annotator2.csv', header=0, names=['row_num','text','gender'], usecols = ['text','gender'])

annotations1 = annotator1['gender'].to_numpy()
annotations2 = annotator2['gender'].to_numpy()

# Calculate Raw Agreement
match = 0
N = len(annotations1)
for i in range(N):
  if annotations1[i] == annotations2[i]:
    match +=1
raw_agreement = (match/N)*100
print(f'Raw Agreement = {raw_agreement}')

# Calculate Cohen's Kappa
kappa = skl.cohen_kappa_score(annotations1, annotations2, labels=['M', 'F'])
print(f'Cohen\'s Kappa = {kappa}')

Raw Agreement = 68.33333333333333
Cohen's Kappa = 0.34104046242774566


###*RAW AGREEMENT*: 68.33 %
###*COHEN'S KAPPA*: 0.34104

## Question 1.6 (10 points):
To form your final dataset, you need to *aggregate* the annotations from both annotators (i.e., for cases where they disagree, you need to choose a single label). Use any method you like other than random label selection to perform this aggregation (e.g., have the two annotators discuss each disagreement and come to consensus, or choose the label you agree with the most). Upload your final dataset to the Colab session (in the same format as the other two files) as final_dataset.csv. Remember to include this file in your final email to us!

### *DESCRIBE YOUR AGGREGATION STRATEGY HERE* ###
I used a third annotated dataset for tie-breaks. This dataset was annotated by another friend. The strategy used for tie-breaks was -  

1. If annotator 1 and 2 are in agreement, keep the annotated class.
2. Else, use the class as annotated in the tie-break data.

In [ ]:
final_dataset = pd.read_csv('final_dataset.csv', header=0, names=['row_num','text','gender'], usecols = ['text','gender'])
print(final_dataset)

                                                  text gender
0    Does seem like a tell that whenever Youngkin g...      M
1    @LisaDTRice Everyone is entitled to voice thei...      F
2    I’m so encouraged that the @POTUS and @VP have...      F
3    Democrats fear for their reelections because o...      F
4    It's SHOCKING to think that the Patriot Act is...      F
..                                                 ...    ...
115  Anyways this one goes out to anyone in any wor...      F
116  Democrats want to pay for their reckless spend...      M
117  Amazon's internal systems sound like a nightma...      M
118  I actually would be surprised if Trump went to...      M
119                      Three words: tax the wealthy.      M

[120 rows x 2 columns]


# Part 2: Text classification

Now we'll move onto fine-tuning  pretrained language models specifically on your dataset. This part of the homework is meant to be an introduction to the HuggingFace library, and it contains code that will potentially be useful for your final projects. Since we're dealing with large models, the first step is to change to a GPU runtime.

## Adding a hardware accelerator

Please go to the menu and add a GPU as follows:

`Edit > Notebook Settings > Hardware accelerator > (GPU)`

Run the following cell to confirm that the GPU is detected.

In [ ]:
import torch

# Confirm that the GPU is detected

assert torch.cuda.is_available()

# Get the GPU device name.
device_name = torch.cuda.get_device_name()
n_gpu = torch.cuda.device_count()
print(f"Found device: {device_name}, n_gpu: {n_gpu}")
device = torch.device("cuda")

Found device: Tesla K80, n_gpu: 1


## Installing Hugging Face's Transformers library
We will use Hugging Face's Transformers (https://github.com/huggingface/transformers), an open-source library that provides general-purpose architectures for natural language understanding and generation with a collection of various pretrained models made by the NLP community. This library will allow us to easily use pretrained models like `BERT` and perform experiments on top of them. We can use these models to solve downstream target tasks, such as text classification, question answering, and sequence labeling.

Run the following cell to install Hugging Face's Transformers library and download a sample data file called tweets.csv that contains tweets about airlines along with a negative, neutral, or positive sentiment rating. Note that you will be asked to link with your Google Drive account to download some of these files. If you're concerned about security risks (there have not been any issues in previous semesters), feel free to make a new Google account and use it for this homework!

In [ ]:
!pip install transformers
!pip install -U -q PyDrive

from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
print('success!')

import os
import zipfile

# Download helper functions file
helper_file = drive.CreateFile({'id': '16HW-z9Y1tM3gZ_vFpJAuwUDohz91Aac-'})
helper_file.GetContentFile('helpers.py')
print('helper file downloaded! (helpers.py)')

# Download sample file of tweets
data_file = drive.CreateFile({'id': '1QcoAmjOYRtsMX7njjQTYooIbJHPc6Ese'})
data_file.GetContentFile('tweets.csv')
print('sample tweets downloaded! (tweets.csv)')

     |████████████████████████████████| 3.1 MB 5.3 MB/s 
     |████████████████████████████████| 895 kB 51.9 MB/s 
     |████████████████████████████████| 3.3 MB 32.1 MB/s 
     |████████████████████████████████| 596 kB 47.6 MB/s 
     |████████████████████████████████| 59 kB 7.1 MB/s 
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13
success!
helper file downloaded! (helpers.py)
sample tweets downloaded! (tweets.csv)


The cell below imports some helper functions we wrote to demonstrate the task on the sample tweet dataset.

In [ ]:
from helpers import tokenize_and_format, flat_accuracy

# Part 1: Data Prep and Model Specifications

Upload your data using the file explorer to the left. We have provided a function below to tokenize and format your data as BERT requires. Make sure that your csv file, titled final_data.csv, has one column "text" and another column "labels" containing integers.

If you run the cell below without modifications, it will run on the tweets.csv example data we have provided. It imports some helper functions we wrote to demonstrate the task on the sample tweet dataset. You should first run all of the following cells with tweets.csv just to see how everything works. Then, once you understand the whole preprocessing / fine-tuning process, change the csv in the below cell to your final_data.csv file, add any extra preprocessing code you wish, and then run the cells again on your own data.

In [104]:
# Upload final_dataset.csv before running

from helpers import tokenize_and_format, flat_accuracy
import pandas as pd

# Pre-processing
df = pd.read_csv('final_dataset.csv', header=0, names=['row_num','text','label'], usecols = ['text','label'])
df['label'].replace(['M', 'F'], [0,1], inplace=True)
#print(df)
#df = pd.read_csv('tweets.csv')

df = df.sample(frac=1).reset_index(drop=True)

texts = df.text.values
labels = df.label.values

### tokenize_and_format() is a helper function provided in helpers.py ###
input_ids, attention_masks = tokenize_and_format(texts)

# Convert the lists into tensors.
input_ids = torch.cat(input_ids, dim=0)
attention_masks = torch.cat(attention_masks, dim=0)
labels = torch.tensor(labels)

# Print sentence 0, now as a list of IDs.
print('Original: ', texts[0])
print('Token IDs:', input_ids[0])

Original:  Great to be with some of the brave men and women of California Law Enforcement! #BackTheBlue
Token IDs: tensor([  101,  2307,  2000,  2022,  2007,  2070,  1997,  1996,  9191,  2273,
         1998,  2308,  1997,  2662,  2375,  7285,   999,  1001,  2067, 10760,
        16558,  5657,   102,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0])


## Create train/test/validation splits

Here we split your dataset into 3 parts: a training set, a validation set, and a testing set. Each item in your dataset will be a 3-tuple containing an input_id tensor, an attention_mask tensor, and a label tensor.



In [105]:

total = len(df)

num_train = int(total * .8)
num_val = int(total * .1)
num_test = total - num_train - num_val

# make lists of 3-tuples (already shuffled the dataframe in cell above)

train_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train)]
val_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train, num_val+num_train)]
test_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_val + num_train, total)]

train_text = [texts[i] for i in range(num_train)]
val_text = [texts[i] for i in range(num_train, num_val+num_train)]
test_text = [texts[i] for i in range(num_val + num_train, total)]


Here we choose the model we want to finetune from https://huggingface.co/transformers/pretrained_models.html. Because the task requires us to label sentences, we wil be using BertForSequenceClassification below. You may see a warning that states that `some weights of the model checkpoint at [model name] were not used when initializing. . .` This warning is expected and means that you should fine-tune your pre-trained model before using it on your downstream task. See [here](https://github.com/huggingface/transformers/issues/5421#issuecomment-652582854) for more info.

In [106]:
from transformers import BertForSequenceClassification, AdamW, BertConfig

model = BertForSequenceClassification.from_pretrained(
    "bert-base-uncased", # Use the 12-layer BERT model, with an uncased vocab.
    num_labels = 2, # The number of output labels.   
    output_attentions = False, # Whether the model returns attentions weights.
    output_hidden_states = False, # Whether the model returns all hidden-states.
)

# Tell pytorch to run this model on the GPU.
model.cuda()


Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.bias']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at

BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0): BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, element

# ACTION REQUIRED #

Define your fine-tuning hyperparameters in the cell below (we have randomly picked some values to start with). We want you to experiment with different configurations to find the one that works best (i.e., highest accuracy) on your validation set. Feel free to also change pretrained models to others available in the HuggingFace library (you'll have to modify the cell above to do this). You might find papers on BERT fine-tuning stability (e.g., [Mosbach et al., ICLR 2021](https://openreview.net/pdf?id=nzpLWnVAyah)) to be of interest.

In [107]:
batch_size = 16
optimizer = AdamW(model.parameters(),
                  lr = 5e-5, # args.learning_rate - default is 5e-5
                  eps = 1e-8 # args.adam_epsilon  - default is 1e-8
                )
epochs = 5

# Fine-tune your model
Here we provide code for fine-tuning your model, monitoring the loss, and checking your validation accuracy. Rerun both of the below cells when you change your hyperparameters above.

In [108]:
batch_size = 16
optimizer = AdamW(model.parameters(),
                  lr = 5e-5, # args.learning_rate - default is 5e-5
                  eps = 1e-8 # args.adam_epsilon  - default is 1e-8
                )
epochs = 10

In [109]:
import numpy as np
# function to get validation accuracy
def get_validation_performance(val_set):
    # Put the model in evaluation mode
    model.eval()

    # Tracking variables 
    total_eval_accuracy = 0
    total_eval_loss = 0

    num_batches = int(len(val_set)/batch_size) + 1

    total_correct = 0

    for i in range(num_batches):

      end_index = min(batch_size * (i+1), len(val_set))

      batch = val_set[i*batch_size:end_index]
      
      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])
      
      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)
        
      # Tell pytorch not to bother with constructing the compute graph during
      # the forward pass, since this is only needed for backprop (training).
      with torch.no_grad():        

        # Forward pass, calculate logit predictions.
        outputs = model(b_input_ids, 
                                token_type_ids=None, 
                                attention_mask=b_input_mask,
                                labels=b_labels)
        loss = outputs.loss
        logits = outputs.logits
            
        # Accumulate the validation loss.
        total_eval_loss += loss.item()
        
        # Move logits and labels to CPU
        logits = logits.detach().cpu().numpy()
        label_ids = b_labels.to('cpu').numpy()

        # Calculate the number of correctly labeled examples in batch
        pred_flat = np.argmax(logits, axis=1).flatten()
        labels_flat = label_ids.flatten()
        num_correct = np.sum(pred_flat == labels_flat)
        total_correct += num_correct
        
    # Report the final accuracy for this validation run.
    avg_val_accuracy = total_correct / len(val_set)
    return avg_val_accuracy



In [110]:
import random

# training loop

# For each epoch...
for epoch_i in range(0, epochs):
    # Perform one full pass over the training set.

    print("")
    print('======== Epoch {:} / {:} ========'.format(epoch_i + 1, epochs))
    print('Training...')

    # Reset the total loss for this epoch.
    total_train_loss = 0

    # Put the model into training mode.
    model.train()

    # For each batch of training data...
    num_batches = int(len(train_set)/batch_size) + 1

    for i in range(num_batches):
      end_index = min(batch_size * (i+1), len(train_set))

      batch = train_set[i*batch_size:end_index]

      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])

      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)

      # Clear the previously calculated gradient
      model.zero_grad()        

      # Perform a forward pass (evaluate the model on this training batch).
      outputs = model(b_input_ids, 
                            token_type_ids=None, 
                            attention_mask=b_input_mask, 
                            labels=b_labels)
      loss = outputs.loss
      logits = outputs.logits

      total_train_loss += loss.item()

      # Perform a backward pass to calculate the gradients.
      loss.backward()

      # Update parameters and take a step using the computed gradient.
      optimizer.step()
        
    # ========================================
    #               Validation
    # ========================================
    # After the completion of each training epoch, measure our performance on
    # our validation set. Implement this function in the cell above.
    print(f"Total loss: {total_train_loss}")
    val_acc = get_validation_performance(val_set)
    print(f"Validation accuracy: {val_acc}")
    
print("")
print("Training complete!")



======== Epoch 1 / 10 ========
Training...
Total loss: 4.495677411556244
Validation accuracy: 0.6666666666666666

======== Epoch 2 / 10 ========
Training...
Total loss: 4.009741008281708
Validation accuracy: 0.4166666666666667

======== Epoch 3 / 10 ========
Training...
Total loss: 3.888410806655884
Validation accuracy: 0.6666666666666666

======== Epoch 4 / 10 ========
Training...
Total loss: 3.3549228608608246
Validation accuracy: 0.6666666666666666

======== Epoch 5 / 10 ========
Training...
Total loss: 2.004633590579033
Validation accuracy: 0.6666666666666666

======== Epoch 6 / 10 ========
Training...
Total loss: 1.249364659190178
Validation accuracy: 0.5

======== Epoch 7 / 10 ========
Training...
Total loss: 0.557344876229763
Validation accuracy: 0.6666666666666666

======== Epoch 8 / 10 ========
Training...
Total loss: 0.47249871492385864
Validation accuracy: 0.75

======== Epoch 9 / 10 ========
Training...
Total loss: 0.13447163812816143
Validation accuracy: 0.666666666666666

# Evaluate your model on the test set
After you're satisfied with your hyperparameters (i.e., you're unable to achieve higher validation accuracy by modifying them further), it's time to evaluate your model on the test set! Run the below cell to compute test set accuracy.


In [111]:
get_validation_performance(test_set)

0.8333333333333334

## Question 2.1 (10 points):
Congratulations! You've now gone through the entire fine-tuning process and created a model for your downstream task. Two more questions left :) First, describe your hyperparameter selection process in words. If you based your process on any research papers or websites, please reference them. Why do you think the hyperparameters you ended up choosing worked better than others? Also, is there a significant discrepancy between your test and validation accuracy? Why do you think this is the case?

### *WRITE YOUR ANSWER HERE*  
I used grid search for hyperparameter selection. Grid search is a straightforward way of selecting hyperparameters, where we go through a list of parameters exhaustively checking all combinations to find the best performing hyperparameters. It is kind of a brute-force approach to hyperparameter tuning.  

I can't say why the chosen hyperparameters are better than the others I used during grid search. Hyperparam tuning is not an exact science, and it could be possible that there are better hyperparams for this model. A better approch for hyperparam tuning would be random search or Bayesian optimization.  

There is a significant discrepancy between my test and validation accuracies. This is probably because the amount of training data is insufficient to generalise well, i.e. the model is unable to learn much from the limited amount of data provided. More training data would be needed to create a model with deep enough insight to generalise well on validation and test set.

## Question 2.2 (20 points):
Finally, perform an *error analysis* on your model. This is good practice for your final project. Write some code in the below code cell to print out the text of up to five test set examples that your model gets **wrong** (if your model gets more than five test examples wrong, randomly choose five of them to print out). Then, in the following text cell, perform a qualitative analysis of these examples. See if you can figure out any reasons for errors that you observe, or if you have any informed guesses (e.g., common linguistic properties of these particular examples). Does this suggest any possible future steps could for your classifier?

In [112]:
## YOUR ERROR ANALYSIS CODE HERE
## print out up to 5 test set examples that your model gets wrong
# Put the model in evaluation mode
model.eval()
#batch_size=16
#print(len(test_set))
input_id_tensors = torch.stack([test[0] for test in test_set])
input_mask_tensors = torch.stack([test[1] for test in test_set])
label_tensors = torch.stack([test[2] for test in test_set])

b_input_ids = input_id_tensors.to(device)
b_input_mask = input_mask_tensors.to(device)
b_labels = label_tensors.to(device)

with torch.no_grad():        
    # Forward pass, calculate logit predictions.
    outputs = model(b_input_ids, 
                            token_type_ids=None, 
                            attention_mask=b_input_mask,
                            labels=b_labels)
    #print(outputs)
    logits = outputs.logits
    
    # Move logits and labels to CPU
    logits = logits.detach().cpu().numpy()
    label_ids = b_labels.to('cpu').numpy()

    # Calculate the number of correctly labeled examples in batch
    pred_flat = np.argmax(logits, axis=1).flatten()
    labels_flat = label_ids.flatten()
    print(pred_flat)
    print(labels_flat)

count = 1
for i in range(len(pred_flat)):
  if pred_flat[i] != labels_flat[i]:
    print(f"Sentence {count}: {test_text[i]}")
    count += 1
    if count > 5:
      break

[1 0 1 0 1 0 0 0 1 0 0 1]
[1 0 0 1 1 0 0 0 1 0 0 1]
Sentence 1: @StellantisNA is poisoning a majority Black neighborhood because it doesn’t care. It only cares about profits. They planted a few trees and called it a day. What a disgrace. Residents warned this would happen &amp; City pushed it through.
Sentence 2: Nine months before Rosa Parks, 15-yr-old Claudette Colvin refused to give up her seat on a bus for a White woman. Now 82, she’s still on probation. @michele_norris


### *DESCRIBE YOUR QUALITATIVE ANALYSIS OF THE ABOVE EXAMPLES HERE*

The task of classifying sentences into genders is (as the annotators noted) extremely vague. Deeper analysis needs to be done to differentiate between male and female writing styles. This may be even more difficult to determine in the case of tweets, where the number of characters is restricetd.  

Both sentences which are being mislabelled in the test set deal with the social issue of race discrimination. Obviously, race is an issue which has no gender, equally affecting both sexes. Which might be why the model would have a hard time classifying these tweets as male or female.



---

Finished? Remember to upload the PDF file of this notebook to Gradescope **AND** email your three dataset files (annotator1.csv, annotator2.csv, and final_data.csv) to cs685instructors@gmail.com with the subject line formatted as **Firstname_Lastname_HW1data**.
